### PR TITLE
chore: add ReadStartingWithUser storage

### PR DESCRIPTION
**Pull request by jon-whit**
## Description
This adds the `ReverseReadTuples` stub to the storage interface to support reverse queries for reverse expansion use cases. This will help us improve the implementation of the ListObjects API by reverse expanding tuples to reveal objects the user has access to.

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [ ] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: storage/test/tuples.go [hunk 1]
```diff
@@ -457,3 +458,59 @@ func TuplePaginationOptionsTest(t *testing.T, datastore storage.OpenFGADatastore
 		}
 	})
 }
+
+func ReverseReadTuplesTest(t *testing.T, datastore storage.OpenFGADatastore) {
+
+	require := require.New(t)
```
Comment (jon-whit) [inline storage/test/tuples.go:464]: This is a stupid simple test. We may want to add more, but this outlines the gist of it.

**Comment by jon-whit** [comment]
Tests will fail until it is implemented. Feel free to help with the implementation for the memory and postgres storage adapters 👍

File: storage/storage.go [hunk 2]
```diff
@@ -230,30 +230,75 @@ type TupleBackend interface {
...
+	) (*openfgapb.Tuple, error)
+
+	ReadUsersetTuples(
+		ctx context.Context,
+		store string,
+		tk *openfgapb.TupleKey,
+	) (TupleIterator, error)
+
+	// ReverseReadTuples performs a reverse read of relationship tuples starting at one or
+	// more user(s) or userset(s) and filtered by object type and relation.
```
Comment (miparnisari) [inline storage/storage.go:268]: The definition of Read API:

> // Read the set of tuples associated with `store` and `key`, which may be partially filled. A key must specify at
>	// least one of `Object` or `User` (or both), and may also optionally constrain by relation.

They both look very similar. 

Why is this one called "reverse"?

File: storage/test/tuples.go [hunk 3]
```diff
@@ -457,3 +458,59 @@ func TuplePaginationOptionsTest(t *testing.T, datastore storage.OpenFGADatastore
...
+	require.NoError(err)
+
+	err = datastore.Write(
+		ctx,
+		storeID,
+		nil,
+		[]*openfgapb.TupleKey{
+			tuple.NewTupleKey("document:doc1", "viewer", "user:jon"),
+			tuple.NewTupleKey("document:doc2", "viewer", "group:eng#member"),
+			tuple.NewTupleKey("document:doc3", "editor", "user:jon"),
```
Comment (miparnisari) [inline storage/test/tuples.go:477]: can we add another tuple here with a type that is not `document`?

File: storage/storage.go (hunk 2 above)
Comment (jon-whit) [reply storage/storage.go:268]: As far as calling it "reverse" I don't think we need to keep that nomenclature if we don't want. But the query is really only intended for query relationship tuples starting at the user/userset and working backwards. It's meant to be indexed as such.

My understanding is that `Read` reads either all tuples for a single object (optionally constrained by a relation) or it reads all objects of any type for a user (optionally constrained by relation).

The goal here is to read all objects that a user or userset has a specific relation with without filtering by object_id whatsoever. As a simple query example,

`SELECT object_type, object_id FROM tuples WHERE store='mystore AND object_type='document' AND relation='viewer' AND user_type=X AND user=Y`

File: storage/test/tuples.go (hunk 3 above)
Comment (jon-whit) [reply storage/test/tuples.go:477]: Yeah, good point.

Comment (jon-whit) [reply storage/test/tuples.go:477]: Updated.

**Comment by jon-whit** [comment]
@miparnisari do you mind renaming this signature based on what we discussed yesterday? I agree that a name such as `ReadStartingWithUser` probably is more logical at this point (until we refactor our storage interface to be more concise).

File: storage/postgres/postgres.go [hunk 4]
```diff
@@ -274,6 +275,28 @@ func (p *Postgres) ReadUsersetTuples(ctx context.Context, store string, tupleKey
 	return &tupleIterator{rows: rows}, nil
 }
 
+func (p *Postgres) ReadStartingWithUser(ctx context.Context, store string, opts storage.ReadStartingWithUserFilter) (storage.TupleIterator, error) {
+	ctx, span := p.tracer.Start(ctx, "postgres.ReadStartingWithUser")
+	defer span.End()
+
+	stmt := "SELECT store, object_type, object_id, relation, _user, ulid, inserted_at FROM tuple WHERE store = $1 AND object_type = $2 AND relation = $3 AND _user = any($4)"
```
Comment (miparnisari) [inline storage/postgres/postgres.go:282]: https://github.com/jackc/pgx/issues/334

Comment (miparnisari) [reply storage/postgres/postgres.go:283]: Also, this query is using the primary key:

```
Index Scan using tuple_pkey on tuple  (cost=0.15..8.17 rows=1 width=200)
"  Index Cond: ((store = '01GCJB1DJ88RJ1KA5TMG5PJXQ4'::text) AND (object_type = 'document'::text) AND (relation = 'owner'::text))"
"  Filter: (_user = ANY ('{maria,jon}'::text[]))"
```

File: storage/memory/memory.go [hunk 5]
```diff
@@ -372,6 +372,40 @@ func (s *MemoryBackend) ReadUsersetTuples(ctx context.Context, store string, key
...
+	defer span.End()
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	var matches []*openfgapb.Tuple
+	for _, t := range s.tuples[store] {
+		if tupleUtils.GetType(t.Key.GetObject()) != filter.ObjectType {
+			continue
+		}
```
Comment (jon-whit) [inline storage/memory/memory.go:390]: Can we add some whitespace between these separate blocks?

File: storage/postgres/postgres.go [hunk 6]
```diff
@@ -275,6 +276,28 @@ func (p *Postgres) ReadUsersetTuples(ctx context.Context, store string, tupleKey
...
+	var targetUsersArg []string
+	for _, u := range opts.UserFilter {
+		targetUser := u.GetObject()
+		if u.GetRelation() != "" {
+			targetUser = strings.Join([]string{u.GetObject(), u.GetRelation()}, "#")
```
Comment (jon-whit) [inline storage/postgres/postgres.go:288]: `fmt.Sprintf("%s#%s", ...)` makes it a little more obvious IMO

File: storage/test/tuples.go [hunk 7]
```diff
@@ -457,3 +458,199 @@ func TuplePaginationOptionsTest(t *testing.T, datastore storage.OpenFGADatastore
...
+					break
+				}
+
+				require.Fail(err.Error())
+			}
+
+			objects = append(objects, tp.GetKey().GetObject())
+		}
+
+		require.Equal([]string{"document:doc1", "document:doc2"}, objects)
```
Comment (jon-whit) [inline storage/test/tuples.go:519]: We don't necessarily guarantee order, so we should sort the two slices (expected and actual). This should be done here and elsewhere.

File: storage/test/tuples.go [hunk 8]
```diff
@@ -457,3 +458,199 @@ func TuplePaginationOptionsTest(t *testing.T, datastore storage.OpenFGADatastore
...
+						Relation: "member",
+					},
+				},
+			},
+		)
+		require.NoError(err)
+
+		var objects []string
+		for {
+			tp, err := iter.Next()
```
Comment (jon-whit) [inline storage/test/tuples.go:507]: This iterator code is used here and below. We should create a helper so we avoid code duplication.

File: storage/test/tuples.go (hunk 7 above)
Comment (miparnisari) [reply storage/test/tuples.go]: Good point!

**Review by jon-whit — COMMENTED** [review]
LGTM! I can't approve because I created it though 😄

**Pull request merged**